Deepintent ID System: add new ID module

sourabhg opened the pull request:
<!--
Thank you for your pull request. Please make sure this PR is scoped to one change, and that any added or changed code includes tests with greater than 80% code coverage. See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [x] Feature
- [x] Does this change affect user-facing APIs or examples documented on http://prebid.org?


## Description of change
New Deepintent user module is added which supports passing of userIds as per deepintent DPES identity.



- contact email of the adapter’s maintainer
sourabh@deepintent.com, prebid@deepintent.com

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/
https://github.com/prebid/prebid.github.io/pull/2830

ChrisHuie commented on modules/.submodules.json:
```diff
@@ -16,7 +16,11 @@
     "zeotapIdPlusIdSystem",
     "haloIdSystem",
     "quantcastIdSystem",
+<<<<<<< HEAD
+    "deepintentDpesIdSystem",
+=======
     "nextrollIdSystem",
+>>>>>>> 81fc1ce9b8471cadb33ebc10a270e8c6d001ea98
```
Can you please explain the concept here. Is it constructing a bit shifting hash?

sourabhg replied:
Hey @ChrisHuie  are you referring to what Deepintent module does ?

sourabhg replied:
I see @ChrisHuie I think i forgot one conflict resolution commit. just pushed it should be good now

ChrisHuie replied:
thanks :)

ChrisHuie commented:
@sourabhg it looks like our circleci tests didn't run on this pr.

sourabhg commented:
@ChrisHuie CircleCI generally runs automatically on all of the things, as I have used the fork flow always.

sourabhg commented:
@ChrisHuie : I just gave you write access to our forked repo. I hope that helps.

ChrisHuie commented:
@sourabhg thanks. Just kicked off the circleci tests. Feel free to remove me from the repo :) I am curious about this situation because I have seen it before but usually with new committers to Prebid. So this is a new situation to me

sourabhg commented:
Thanks @ChrisHuie May I know when this can be merged ?

smenzer commented on modules/deepintentDpesIdSystem.js:
```diff
@@ -0,0 +1,80 @@
+      configStorage.name &&
+      configStorage.type === 'cookie'
+    ) {
+      id = readCookie(configStorage.name);
```
why are you manually reading from the cookie if you are asking publishers to set the `storage` config?  the userId module will already handle storage and retrieval you just need to add the `cacheIdObj` to the `getId()` signature

sourabhg replied:
This code was started when cacheIdObj signature was not there, the code will be later changed to accomodate this in phase 2.

sourabhg replied:
As a total change I will be changing the description in Doc and adding unit test cases in eids_spec.js and userId_spec.js thanks for letting me know.

smenzer replied:
the `storedId` has been part of the signature for years now...

patmmccann replied:
Why wait for a phase two,  I'm in the do it right the first time camp

sourabhg replied:
storage will not be handled by UserModule as we do not use any http calls to be made to retrieve id from server and store it via user module. This id is stored by our select publishers side which needs to register on our platform and thus we have to read from storage manually, thus keeping the structure same as per the plan.

sourabhg replied:
changes are completed as suggested.

smenzer commented on modules/deepintentDpesIdSystem.md:
```diff
@@ -0,0 +1,49 @@
+# Deepintent DPES ID
+
+The Deepintent DPES ID or Healthcare Identity is a shared, neutral identifier that publishers and ad tech platforms can use to recognise users even in environments where 3rd party cookies are not available. The Deepintent DPES ID is designed to respect users' privacy choices and publishers’ preferences throughout the advertising value chain. 
```
This is an exact copy of the [ID5 description](https://github.com/prebid/Prebid.js/blob/master/modules/id5IdSystem.md#L3). You should not just be copying and pasting here, it's meant to describe YOUR id solution.

sourabhg replied:
This is exactly what our solution does as well and describes what we do. If you still persist I can change it with diffrent language

patmmccann replied:
run that by your corporate pr folks, they may frown on plaigarism

sourabhg replied:
Thank you changed the language now.

smenzer commented on modules/deepintentDpesIdSystem.md:
```diff
@@ -0,0 +1,49 @@
+      },
+      storage: {
+        type: 'cookie',           // "html5" is the required storage type option is "html5"
+        name: '_di',            // change to the cookie name you wish to read from, optional if storage type is html5
```
why do you treat cookie and html5 differently?

sourabhg replied:
We give a way to specify cookie name for pubs and for localstorage key is already set with us. This is to accomodate the exisitng solution we have. Once we made changes to that code we might change the working here as well for now no changes and thank you for suggestion.

sourabhg replied:
Changes are completed as suggested.

smenzer commented on modules/userId/eids.js:
```diff
@@ -207,7 +207,10 @@ const USER_IDS_CONFIG = {
-
+  'deepintentId': {
+    source: 'deepintent.com',
+    atype: 1
```
if this is an email-based identifier, then it should be `3` here I think. @patmmccann you've been reviewing a lot of these, do you agree?

sourabhg replied:
I will make the change, Is there a list as well as which is being referred ?

patmmccann replied:
https://github.com/InteractiveAdvertisingBureau/AdCOM/blob/master/AdCOM%20v1.0%20FINAL.md#list_agenttypes

smenzer reviewed (changes requested):
A few comments in the code. You also need to add tests in the `userId_spec.js` and `eids_spec.js` files

patmmccann commented:
Thank you for fixing the atype; everyone just put 1 at first and I've been trying to make them accurate.

sourabhg commented:
@smenzer @patmmccann : Added in all the necessery test cases, changed the documentation, changed the atype to 3 and also changed the logic to pick from cacheIdObject. 

@ChrisHuie : somehow CircleCI again doesnt work through my commits can you help. ?

smenzer commented on modules/deepintentDpesIdSystem.js:
```diff
@@ -0,0 +1,51 @@
+   * @returns {{deepintentId:Object}}
+   */
+  decode(value, config) {
+    if (value) {
+      return {
+        'deepintentId': value
+      }
+    } else {
+      return undefined
+    }
```
this could be simplified with 

```suggestion
    return value ? { 'deepintentId': value } : undefined;
```

sourabhg replied:
Made the change.

smenzer commented on modules/deepintentDpesIdSystem.md:
```diff
@@ -0,0 +1,42 @@
+The Deepintent DPES ID is free to use, but requires a simple registration with Deepintent. Please contact your sales rep to get started
+
+
+## Deepintent DPES ID Configuration
```
as mentioned in the docs PR, it seems to be required that publishers have additional Deepintent code on their page to retrieve and store the DPES ID in order for the user id module to pick it up. Assuming that's the case, I think that should be clearly documented here somewhere

sourabhg replied:
@smenzer : as suggested I added more details about the Deepintent registration and a tag code that publisher needs to integrate to get started with this user module, can we merge this now ?

smenzer reviewed (changes requested):
Just needs clarification on additional code required to work, but otherwise this is good to go

sourabhg commented:
@smenzer @ChrisHuie : All review comments have been addressed can we approve and merge this now ?

smenzer reviewed (approved):
Thanks for the updates...LGTM
